lib/hooks/index.js

```diff
@@ -97,25 +67,51 @@ function findModuleVersion(request, parent, load) {
     if (mainScriptDir.indexOf(paths[i]) === 0) {
       // This is unsafe on windows if request contains a namespace
       // TODO: convert path separators in request
-      var pjson = load(path.join(paths[i], request, 'package.json'));
-      return pjson.version;
+      return path.join(paths[i], request);
     }
   }
   return null;
 }
 
+/**
+ * Determines the version of the module located at `modulePath`.
+ *
+ * @param {?string} modulePath The absolute path to the root directory of the
+ *    module being loaded. This may be null if we are loading an internal module
+ *    such as http.
+ */
+function findModuleVersion(modulePath, load) {
+  return modulePath ? load(path.join(modulePath, 'package.json')).version :
+    process.version;
+}
+
 function activate(agent) {
 
   logger = agent.logger;
 
-  function loadHookAndPatch(loadFn, instrumentation) {
-    assert(instrumentation.module);
-    assert(!instrumentation.hook);
+  function loadHookAndPatch(loadFn, instrumentation, moduleRoot, version) {
+    var modulePatch = instrumentation.patches[moduleRoot];
+    if (!modulePatch) {
+      // Load the hook. This file, i.e. index.js, becomes the parent module.
+      var moduleHook = loadFn(instrumentation.file, module, false);
+      modulePatch = moduleHook(version, agent);
+    }
+    Object.keys(modulePatch).forEach(function(file) {
+      if (!modulePatch[file].module) {
+        var loadPath = moduleRoot ? path.join(moduleRoot, file) : file;
+        modulePatch[file].module = loadFn(loadPath, module, false);
+      }
+      modulePatch[file].patch(modulePatch[file].module);
+      modulePatch[file].active = true;
+    });
+    instrumentation.patches[moduleRoot] = modulePatch;
+  }
 
-    // Load the hook. This file, i.e. index.js, becomes the parent module.
-    var moduleHook = loadFn(instrumentation.file, module, false);
-    instrumentation.hook = moduleHook;
-    moduleHook.patch(instrumentation.module, agent, instrumentation.version);
+  function moduleAlreadyPatched(instrumentation, moduleRoot) {
+    var modulePatch = instrumentation.patches[moduleRoot];
+    return modulePatch && Object.keys(modulePatch).every(function(curr) {
```

Comment:
This is not currently possible but there isn't a nice way to store this info a level up without adding anther layer of wrapping around things. Additionally, there isn't anything fundamental to say that they couldn't be patched individually in the future.